File: benchmarks/primes_sieve/primes_sieve.py
import sys
import time

n = int(sys.argv[1]) if len(sys.argv) > 1 else 1000000

start = time.perf_counter()

sieve = [False] * (n + 1)

i = 2
while i * i <= n:
    if not sieve[i]:
        j = i * i
        while j <= n:
            sieve[j] = True
            j += i
    i += 1

count = 0
for i in range(2, n + 1):
    if not sieve[i]:
        count += 1

elapsed = (time.perf_counter() - start) * 1000
print(f"TIME_MS:{elapsed:.2f}", file=sys.stderr)
print(count)

File: benchmarks/primes_sieve/primes_sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>

int main(int argc, char *argv[]) {
    int n = argc > 1 ? atoi(argv[1]) : 1000000;

    struct timeval start, end;
    gettimeofday(&start, NULL);

    char *sieve = calloc(n + 1, 1);

    for (int i = 2; i * i <= n; i++) {
        if (!sieve[i]) {
            for (int j = i * i; j <= n; j += i) {
                sieve[j] = 1;
            }
        }
    }

    int count = 0;
    for (int i = 2; i <= n; i++) {
        if (!sieve[i]) count++;
    }

    gettimeofday(&end, NULL);
    double elapsed_ms = (end.tv_sec - start.tv_sec) * 1000.0 + (end.tv_usec - start.tv_usec) / 1000.0;
    fprintf(stderr, "TIME_MS:%.2f\n", elapsed_ms);

    printf("%d\n", count);
    free(sieve);
    return 0;
}
